### ISSUE TITLE
Where is the C Callable Library?

### ISSUE THREAD

**Issue by scottconstable** [coin-or/SYMPHONY#172]
The SYMPHONY manual has a section describing a "C Callable Library API." I have installed the Debian development package for SYMPHONY, and the installed libraries libSym.(a|so) only export mangled C++ names. I also build SYMPHONY from source, and observe the same thing: all of the compiled binaries export C++-style symbols. So is this a documentation bug? Or is there another package with C APIs that I am missing?

**Comment by tkralphs** [comment]
Good point. You could say it's a documentation bug, but SYMPHONY is written in pure C and the API is in C. Once it became integrated into the COIN-OR ecosystem, we started to build it in C++ because the rest of its dependencies are in C++ and the COIN-OR build system builds C++ by default. That's why the Debian package is exporting C++ symbols. I guess if we just wrap the header in `extern "C"`, that should be all that's required. I will give it a try. What are you trying to do that requires a C interface?

**Comment by tkralphs** [comment]
I just wrapped `symphony.h` in `extern "C"` and that seemed to do the trick. I don't see any reason not to just do that in the next version, actually, thanks for pointing this out. In the meantime, it's straightforward to build it in Debian/Ubuntu (see [INSTALL](https://github.com/coin-or/SYMPHONY/blob/master/INSTALL.md)). If you point `PKG_CONFIG_PATH` to `/usr/lib/x86_64-linux-gnu/pkgconfig`, then it will link to existing installed Debian libraries. 

So just wrap `symphony.h` in `extern "C"{ }` and then do
```
export PKG_CONFIG_PATH=/usr/lib/x86_64-linux-gnu/pkgconfig
wget https://raw.githubusercontent.com/coin-or/coinbrew/master/coinbrew
chmod u+x coinbrew
./coinbrew fetch SYMPHONY
./coinbrew build SYMPHONY --no-prompt --skip-dependencies
```
Then `dist/lib` should contain a library with a C API.

**Comment by scottconstable** [comment]
IMO the standard way to export an interface that should be callable by either C or C++ code is to wrap the interface with:
```C
#ifdef __cplusplus
extern "C" {
#endif

// C/C++ API

#ifdef __cplusplus
} // end extern "C"
#endif
```
This way, none of the symbol names within the block is mangled (either by client code, or the library itself), and either C or C++ code can link to the APIs without any additional effort. This is how other callable libraries such as GLPK support both C and C++.

Just beware that if you upstream this change, any clients using libSym will need to recompile and relink, because the symbol names will have changed.

**Comment by tkralphs** [comment]
Right, yes, the `#ifdef __cplusplus` is a good idea. I'll probably only introduce this when releasing a new stable version, which should hopefully eliminate any surprises. Thanks!

**Comment by tkralphs** [comment]
Fixed via de97de6b.
